src/meta/src/barrier/mod.rs

```diff
@@ -254,80 +382,143 @@ where
+                    env.clone(),
+                    metrics,
+                    &command_ctx,
+                    heap.clone(),
+                    hummock_manager,
+                    wait_recovery,
+                    wait_rx,
+                )
+                .await;
+
+                // debug!("over,new_epoch{:?}",new_epoch);
+                match result {
+                    Ok(responses) => {
+                        // Notify about collected first.
+                        notifiers.iter_mut().for_each(Notifier::notify_collected);
+                        // Then try to finish the barrier for Create MVs.
+                        let actors_to_finish = command_ctx.actors_to_finish();
+                        barrier_send_tx_clone
+                            .send(BarrierSendResult::Ok {
+                                new_epoch,
+                                actors_to_finish,
+                                notifiers,
+                                responses,
+                            })
+                            .unwrap();
                     }
-
-                    state.prev_epoch = new_epoch;
-                }
-                Err(e) => {
-                    notifiers
-                        .into_iter()
-                        .for_each(|notifier| notifier.notify_collection_failed(e.clone()));
-                    if self.enable_recovery {
-                        // If failed, enter recovery mode.
-                        let (new_epoch, actors_to_finish, finished_create_mviews) =
-                            self.recovery(new_epoch).await;
-                        unfinished = UnfinishedNotifiers::default();
-                        unfinished.add(new_epoch.0, actors_to_finish, vec![]);
-                        for finished in finished_create_mviews {
-                            unfinished.finish_actors(finished.epoch, once(finished.actor_id));
+                    Err(err_msg) => {
+                        notifiers.into_iter().for_each(|notifier| {
+                            notifier.notify_collection_failed(err_msg.clone())
+                        });
+                        let mut is_over = false;
+                        if heap.read().await.len() == 1 {
+                            is_over = true;
                         }
-
-                        state.prev_epoch = new_epoch;
-                    } else {
-                        panic!("failed to execute barrier: {:?}", e);
+                        barrier_send_tx_clone
+                            .send(BarrierSendResult::Err {
+                                new_epoch,
+                                err_msg,
+                                is_over,
+                            })
+                            .unwrap();
                     }
                 }
-            }
-
-            state.update(self.env.meta_store()).await.unwrap();
+                heap.write().await.pop();
```

Comment:
Originally I don't get why we pop the heap before checking the peek and later on I realized that in `run_inner`, we wait for the epoch to be the head of the heap so we need to pop here. Can you refactor the codes to make it clearer.

Reply:
OkOk , I think I need to make a big change.